File: lib/ml.js
// lib/ml.js
// Small, from-scratch ML utilities — no external dependencies. Kept
// intentionally simple and auditable rather than pulling in a heavy ML
// library for a personal site running on Node serverless functions.
//
// Only the Race Time Predictor's math lives here now — the workout-type
// clustering model (k-means) was removed after several rounds of tuning
// still weren't classifying real training sessions well enough.

// ---------- Power-law regression: time = a * distance^b ----------
// This is a log-log linear regression (least squares), used to fit a
// PERSONALIZED version of the Riegel race-prediction formula
// (generic Riegel uses a fixed exponent b≈1.06 for everyone; here b is
// fitted from the athlete's own PR curve).
function powerLawFit(points) {
  const n = points.length;
  const xs = points.map(p => Math.log(p.x));
  const ys = points.map(p => Math.log(p.y));
  const xMean = xs.reduce((a, b) => a + b, 0) / n;
  const yMean = ys.reduce((a, b) => a + b, 0) / n;
  let num = 0, den = 0;
  for (let i = 0; i < n; i++) { num += (xs[i] - xMean) * (ys[i] - yMean); den += (xs[i] - xMean) ** 2; }
  const b = den ? num / den : 1.06;
  const logA = yMean - b * xMean;
  const a = Math.exp(logA);
  let ssRes = 0, ssTot = 0;
  for (let i = 0; i < n; i++) {
    const yPred = logA + b * xs[i];
    ssRes += (ys[i] - yPred) ** 2;
    ssTot += (ys[i] - yMean) ** 2;
  }
  const r2 = ssTot ? 1 - ssRes / ssTot : 1;
  return { a, b, r2 };
}
function predictTime(model, distanceKm) { return model.a * Math.pow(distanceKm, model.b); }

function mean(arr) { return arr.reduce((a, b) => a + b, 0) / arr.length; }

// ---------- Leave-one-out cross-validation for the power-law fit ----------
// Refits the curve n times (once per training point held out) to get an
// honest estimate of prediction error — not just how well the curve fits
// the same points it was trained on.
function loocvPowerLaw(points) {
  const n = points.length;
  if (n < 4) return null; // need at least 3 points left over each fold to fit
  let sumAbsErrS = 0, sumAbsPct = 0;
  for (let i = 0; i < n; i++) {
    const rest = points.filter((_, j) => j !== i);
    const curve = powerLawFit(rest);
    const predicted = predictTime(curve, points[i].x);
    const actual = points[i].y;
    sumAbsErrS += Math.abs(predicted - actual);
    sumAbsPct += Math.abs(predicted - actual) / actual;
  }
  return {
    maeSeconds: Math.round(sumAbsErrS / n),
    mapePct: Math.round((sumAbsPct / n) * 1000) / 10 // Mean Absolute Percentage Error
  };
}

module.exports = { powerLawFit, predictTime, mean, loocvPowerLaw };

File: lib_py/ml.py
# lib_py/ml.py
# Python port of lib/ml.js's race-prediction math — same algorithm
# (log-log linear regression / personalized Riegel formula), but using
# numpy + scikit-learn instead of hand-rolled least squares, so the two
# implementations can be compared side by side.

import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import LeaveOneOut


def power_law_fit(points):
    """Fit time = a * distance^b via log-log linear regression.

    points: list of (distance_km, time_seconds) tuples.
    Returns (a, b, r2).
    """
    x = np.array([[np.log(p[0])] for p in points])
    y = np.array([np.log(p[1]) for p in points])

    model = LinearRegression().fit(x, y)
    b = float(model.coef_[0])
    log_a = float(model.intercept_)
    a = float(np.exp(log_a))
    r2 = float(model.score(x, y))
    return a, b, r2


def predict_time(a, b, distance_km):
    return a * (distance_km ** b)


def loocv_power_law(points):
    """Leave-one-out cross-validation: refit the curve once per held-out
    point to get an honest error estimate, not just goodness-of-fit on the
    same data the curve was trained on.
    """
    n = len(points)
    if n < 4:
        return None

    x_all = np.array([[np.log(p[0])] for p in points])
    y_all = np.array([np.log(p[1]) for p in points])
    distances = np.array([p[0] for p in points])
    times = np.array([p[1] for p in points])

    loo = LeaveOneOut()
    abs_errors = []
    abs_pct_errors = []

    for train_idx, test_idx in loo.split(x_all):
        model = LinearRegression().fit(x_all[train_idx], y_all[train_idx])
        b = model.coef_[0]
        a = np.exp(model.intercept_)
        i = test_idx[0]
        predicted = a * (distances[i] ** b)
        actual = times[i]
        abs_errors.append(abs(predicted - actual))
        abs_pct_errors.append(abs(predicted - actual) / actual)

    mae_seconds = round(float(np.mean(abs_errors)))
    mape_pct = round(float(np.mean(abs_pct_errors)) * 1000) / 10
    return {"maeSeconds": mae_seconds, "mapePct": mape_pct}


def mean(values):
    return sum(values) / len(values) if values else 0
